File: cpp/mpi_pi_sr.c
// This program is to caculate PI using MPI and numerical integration
// If f(x) = 4 * sqrt(1 - x^2), then PI is the intergral of f(x) from 0 to 1

#include <cmath>
#include <mpi.h>

constexpr double f(double x)
{
    return sqrt(1 - x * x);
}
 
int main(int argc, char* argv[])
{
    constexpr int64_t N = 1e7;

    constexpr double from = 0.;
    constexpr double to = 1.;
    
    constexpr double d = (to - from) / static_cast<double>(N);

    // Initialize MPI
    MPI_Init(&argc, &argv);

    // Get process id and total number of processes size
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);    
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    
    // Wait until all processes arrive here, then continue
    MPI_Barrier(MPI_COMM_WORLD);
    const double begin = MPI_Wtime();
    
    //Each process caculates a part of the sum. 
    double subsum = 0;
    for (int64_t i = rank; i < N; i += size)
    {
        const double x = from + i * d;
        const double fx = f(x);
        subsum += fx * d;
    }
    
    double sum = subsum;
    //Sum up all sub sums
    if (rank)
        MPI_Send(&subsum, 1, MPI_DOUBLE, 0, 0, MPI_COMM_WORLD);
    else {
        for (int i = 1; i < size; i++) {
            double subsum2;
            MPI_Recv(&subsum2, 1, MPI_DOUBLE, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, 0);
            sum += subsum2;
        }
    }
    
    //Synchronize all processes and get the end time
    MPI_Barrier(MPI_COMM_WORLD);
    const double end = MPI_Wtime();
    
     //Caculate and print PI
    if (!rank)
        std::cout << "Processes = " << size << ", Exec. duration = " << end - begin << "s, PI = " << 4 * sum << ", error = "  << 4 * sum - M_PI << std::endl;
    
    MPI_Finalize();
    
    return 0;
}

File: python/mpi_pi_sr.py
#!/usr/bin/env python 

# Caculate PI using MPI and numerical integration
# of a quarter circle

from mpi4py import MPI
import math

def f(x: float) -> float:
	return math.sqrt(1 - x ** 2)

N: int = int(1e6)

fromV: float = 0.
toV: float = 1.

d: float = (toV - fromV) / N

# Initialize MPI
comm = MPI.COMM_WORLD
size: int = comm.Get_size()
rank: int = comm.Get_rank()

# Wait until all processes arrive here, then continue
comm.Barrier()
begin: float = MPI.Wtime()

# Each process caculates a part of the sum. 
subsum: float = 0
for i in range(rank, N, size):
    x: float = fromV + i * d
    fx: float = f(x)
    subsum += fx * d;

# Sum up all sub sums
sum: float = subsum
if rank != 0:
    comm.send(subsum, dest=0, tag=rank)
else:
    for i in range(1, size):
        subsum2 = comm.recv()
        sum += subsum2;

# Synchronize all processes and get the end time
comm.Barrier()
end: float = MPI.Wtime()

# Caculate and print PI
if rank == 0:
	print("Processes = ", size, ", Exec. duration = ", end - begin, "s, PI = ", 4 * sum, ", error = ", 4 * sum - math.pi)


MPI.Finalize()